src/Chainweb/BlockHeader.hs

```diff
@@ -374,16 +378,40 @@ epochStart ph@(ParentHeader p) adj (BlockCreationTime bt)
     --    b) adjusting timeBlocked in both directions, possibly based off the
     --       avg?
     --
-    timeBlocked = maximum adjCreationTimes .-. _blockCreationTime p
+    _adjustmentMax = maximum adjCreationTimes .-. _blockCreationTime p
         -- the maximum is at least @_blockCreationTime p@ and thus the result is
         -- greater or equal 0.
 
+    -- This computes @mean adjCreationTimes - _blockCreationTime p
+    --
+    -- It holds that
+    --
+    -- \(\left(mean_{0 \leq i < n} a_i\right)
+    -- = \frac{\sum_{0 \leq i < n} a_i}{n} - t
+    -- = \frac{\left(sum_{0 \leq i < n} a_i\right) - \left(\sum_{0 \leq i < n} t\right)}{n}
+    -- = \frac{sum_{0 \leq i < n} (a_i - t)}{n}
+    -- \)
+    --
+    --
+    -- this is numberically sound because we compute the differences on integral
+    -- types without rounding.
+    --
+    _adjustmentAvg = x `divTimeSpan` length adjCreationTimes
+      where
+        x :: TimeSpan Micros
+        x = foldr1 addTimeSpan $ (.-. _blockCreationTime p) <$> adjCreationTimes
+
+    -- This includes the parent header itself.
     adjCreationTimes = _blockCreationTime . _parentHeader <$> HM.insert cid ph adj
 
     parentIsFirstOnNewChain
         = _blockHeight p > 1 && _blockHeight p == genesisHeight ver cid + 1
 {-# INLINE epochStart #-}
 
+divTimeSpan :: Integral a => (TimeSpan a) -> Int -> TimeSpan a
```

Comment:
why not just take an `a` here. Soemthing special about `Int`?

Reply:
I guess, it could be any integral. This should be moved to `Chainweb.Time` anyways.